File: src/lib/engine/normalization.ts
// CNC Cut Plan Engine — Tree Normalization (Canonical Hierarchy Rebuilding)

import { TreeNode } from './types';
import { gid } from './tree-utils';

interface AbsRect {
  x: number;
  y: number;
  w: number;
  h: number;
  label?: string;
}

function extractAbsoluteRects(tree: TreeNode, usableW: number, usableH: number): AbsRect[] {
  const rects: AbsRect[] = [];
  const T = tree.transposed || false;

  let xOff = 0;
  for (const colX of tree.filhos) {
    for (let ix = 0; ix < colX.multi; ix++) {
      let yOff = 0;
      for (const yNode of colX.filhos) {
        for (let iy = 0; iy < yNode.multi; iy++) {
          let zOff = 0;
          for (const zNode of yNode.filhos) {
            for (let iz = 0; iz < zNode.multi; iz++) {
              if (zNode.filhos.length === 0) {
                if (T) {
                  rects.push({ x: yOff, y: xOff, w: yNode.valor, h: zNode.valor, label: zNode.label });
                } else {
                  rects.push({ x: xOff + zOff, y: yOff, w: zNode.valor, h: yNode.valor, label: zNode.label });
                }
              } else {
                let wOff = 0;
                for (const wNode of zNode.filhos) {
                  for (let iw = 0; iw < wNode.multi; iw++) {
                    if (wNode.filhos.length === 0) {
                      if (T) {
                        rects.push({ x: yOff + wOff, y: xOff + zOff, w: wNode.valor, h: zNode.valor, label: wNode.label });
                      } else {
                        rects.push({ x: xOff + zOff, y: yOff + wOff, w: zNode.valor, h: wNode.valor, label: wNode.label });
                      }
                    } else {
                      let qOff = 0;
                      for (const qNode of wNode.filhos) {
                        for (let iq = 0; iq < qNode.multi; iq++) {
                          if (qNode.filhos.length === 0) {
                            if (T) {
                              rects.push({ x: yOff + wOff, y: xOff + zOff + qOff, w: wNode.valor, h: qNode.valor, label: qNode.label });
                            } else {
                              rects.push({ x: xOff + zOff + qOff, y: yOff + wOff, w: qNode.valor, h: wNode.valor, label: qNode.label });
                            }
                          } else {
                            let rOff = 0;
                            for (const rNode of qNode.filhos) {
                              for (let ir = 0; ir < rNode.multi; ir++) {
                                if (T) {
                                  rects.push({ x: yOff + wOff + rOff, y: xOff + zOff + qOff, w: rNode.valor, h: qNode.valor, label: rNode.label });
                                } else {
                                  rects.push({ x: xOff + zOff + qOff, y: yOff + wOff + rOff, w: qNode.valor, h: rNode.valor, label: rNode.label });
                                }
                                rOff += rNode.valor;
                              }
                            }
                          }
                          qOff += qNode.valor;
                        }
                      }
                    }
                    wOff += wNode.valor;
                  }
                }
              }
              zOff += zNode.valor;
            }
          }
          yOff += yNode.valor;
        }
      }
      xOff += colX.valor;
    }
  }

  return rects;
}

function enforceCutMinBreak(cuts: number[], bound: number, minBreak: number): number[] {
  if (minBreak <= 0 || cuts.length === 0) return cuts;
  let filtered = [...cuts].sort((a, b) => a - b);
  let changed = true;
  while (changed) {
    changed = false;
    const boundaries = [0, ...filtered, bound];
    for (let i = 0; i < boundaries.length - 1; i++) {
      const gap = boundaries[i + 1] - boundaries[i];
      if (gap > 0 && gap < minBreak) {
        // Remove whichever interior boundary creates the offending small gap.
        // Prefer removing the right boundary; if it's `bound`, remove the left.
        if (boundaries[i + 1] < bound) {
          filtered = filtered.filter(c => c !== boundaries[i + 1]);
        } else {
          filtered = filtered.filter(c => c !== boundaries[i]);
        }
        changed = true;
        break;
      }
    }
  }
  return filtered;
}

function findVerticalCuts(rects: AbsRect[], bx: number, by: number, bw: number, bh: number, minBreak = 0): number[] {
  const edges = new Set<number>();
  for (const r of rects) {
    const left = r.x - bx;
    const right = r.x + r.w - bx;
    if (left > 0.5 && left < bw - 0.5) edges.add(Math.round(left));
    if (right > 0.5 && right < bw - 0.5) edges.add(Math.round(right));
  }

  const validCuts: number[] = [];
  for (const cx of [...edges].sort((a, b) => a - b)) {
    const absCx = bx + cx;
    const straddles = rects.some(r => r.x < absCx - 0.5 && r.x + r.w > absCx + 0.5);
    if (!straddles) validCuts.push(cx);
  }

  return enforceCutMinBreak(validCuts, Math.round(bw), minBreak);
}

function findHorizontalCuts(rects: AbsRect[], bx: number, by: number, bw: number, bh: number, minBreak = 0): number[] {
  const edges = new Set<number>();
  for (const r of rects) {
    const bottom = r.y - by;
    const top = r.y + r.h - by;
    if (bottom > 0.5 && bottom < bh - 0.5) edges.add(Math.round(bottom));
    if (top > 0.5 && top < bh - 0.5) edges.add(Math.round(top));
  }

  const validCuts: number[] = [];
  for (const cy of [...edges].sort((a, b) => a - b)) {
    const absCy = by + cy;
    const straddles = rects.some(r => r.y < absCy - 0.5 && r.y + r.h > absCy + 0.5);
    if (!straddles) validCuts.push(cy);
  }

  return enforceCutMinBreak(validCuts, Math.round(bh), minBreak);
}

function rectsInBounds(rects: AbsRect[], bx: number, by: number, bw: number, bh: number): AbsRect[] {
  return rects.filter(r =>
    r.x >= bx - 0.5 && r.x + r.w <= bx + bw + 0.5 &&
    r.y >= by - 0.5 && r.y + r.h <= by + bh + 0.5
  );
}

function buildCanonicalTree(rects: AbsRect[], usableW: number, usableH: number, minBreak = 0): TreeNode {
  const root: TreeNode = { id: 'root', tipo: 'ROOT', valor: usableW, multi: 1, filhos: [] };

  if (rects.length === 0) return root;

  type Level = 'X' | 'Y' | 'Z' | 'W' | 'Q' | 'R';
  const levelSequence: Level[] = ['X', 'Y', 'Z', 'W', 'Q', 'R'];
  const isVertical = (level: Level) => level === 'X' || level === 'Z' || level === 'Q';

  function subdivide(
    parentNode: TreeNode,
    levelIdx: number,
    subRects: AbsRect[],
    bx: number, by: number, bw: number, bh: number,
  ): void {
    if (subRects.length === 0 || levelIdx >= levelSequence.length) return;

    const level = levelSequence[levelIdx];
    const vertical = isVertical(level);

    const cuts = vertical
      ? findVerticalCuts(subRects, bx, by, bw, bh, minBreak)
      : findHorizontalCuts(subRects, bx, by, bw, bh, minBreak);

    if (cuts.length === 0) {
      if (subRects.length === 1) {
        const r = subRects[0];
        // Snap dimension to parent bound when residual < minBreak to avoid invalid small cuts.
        const snapW = (minBreak > 0 && bw - r.w > 0 && bw - r.w < minBreak) ? bw : r.w;
        const snapH = (minBreak > 0 && bh - r.h > 0 && bh - r.h < minBreak) ? bh : r.h;
        if (vertical) {
          const node: TreeNode = { id: gid(), tipo: level, valor: Math.round(snapW), multi: 1, filhos: [], label: r.label };
          parentNode.filhos.push(node);
          if (levelIdx + 1 < levelSequence.length) {
            const nextLevel = levelSequence[levelIdx + 1];
            const hNode: TreeNode = { id: gid(), tipo: nextLevel, valor: Math.round(snapH), multi: 1, filhos: [], label: r.label };
            node.filhos.push(hNode);
          }
        } else {
          const node: TreeNode = { id: gid(), tipo: level, valor: Math.round(snapH), multi: 1, filhos: [], label: r.label };
          parentNode.filhos.push(node);
          if (levelIdx + 1 < levelSequence.length) {
            const nextLevel = levelSequence[levelIdx + 1];
            const wNode: TreeNode = { id: gid(), tipo: nextLevel, valor: Math.round(snapW), multi: 1, filhos: [], label: r.label };
            node.filhos.push(wNode);
          }
        }
        return;
      }

      // At the X level with no valid vertical cuts: preserve canonical ROOT→X→Y→… structure
      // by wrapping all rects in a single full-width X node. Without this, Y nodes become
      // direct children of ROOT, breaking the assumed X→Y→Z→W→Q traversal order everywhere.
      if (level === 'X') {
        const xNode: TreeNode = { id: gid(), tipo: 'X', valor: Math.round(bw), multi: 1, filhos: [] };
        subdivide(xNode, levelIdx + 1, subRects, bx, by, bw, bh);
        parentNode.filhos.push(xNode);
        return;
      }

      subdivide(parentNode, levelIdx + 1, subRects, bx, by, bw, bh);
      return;
    }

    const boundaries = vertical
      ? [0, ...cuts, Math.round(bw)]
      : [0, ...cuts, Math.round(bh)];

    const uniqueBounds = [...new Set(boundaries)].sort((a, b) => a - b);

    for (let i = 0; i < uniqueBounds.length - 1; i++) {
      const segStart = uniqueBounds[i];
      const segEnd = uniqueBounds[i + 1];
      const segSize = segEnd - segStart;
      if (segSize < 1) continue;

      let segBx: number, segBy: number, segBw: number, segBh: number;
      if (vertical) {
        segBx = bx + segStart;
        segBy = by;
        segBw = segSize;
        segBh = bh;
      } else {
        segBx = bx;
        segBy = by + segStart;
        segBw = bw;
        segBh = segSize;
      }

      const segRects = rectsInBounds(subRects, segBx, segBy, segBw, segBh);

      if (segRects.length === 0) continue;

      const nodeValor = vertical ? segBw : segBh;
      const node: TreeNode = { id: gid(), tipo: level, valor: Math.round(nodeValor), multi: 1, filhos: [] };

      if (segRects.length === 1) {
        const r = segRects[0];
        const fillsW = Math.abs(r.w - segBw) < 1;
        const fillsH = Math.abs(r.h - segBh) < 1;

        if (fillsW && fillsH) {
          node.label = r.label;
          if (levelIdx + 1 < levelSequence.length) {
            const nextLevel = levelSequence[levelIdx + 1];
            const childValor = vertical ? Math.round(segBh) : Math.round(segBw);
            const child: TreeNode = { id: gid(), tipo: nextLevel, valor: childValor, multi: 1, filhos: [], label: r.label };
            node.filhos.push(child);
          }
          parentNode.filhos.push(node);
          continue;
        }
      }

      subdivide(node, levelIdx + 1, segRects, segBx, segBy, segBw, segBh);
      parentNode.filhos.push(node);
    }
  }

  subdivide(root, 0, rects, 0, 0, usableW, usableH);
  return root;
}

function deepCloneWithNewIds(node: TreeNode): TreeNode {
  return {
    id: gid(),
    tipo: node.tipo,
    valor: node.valor,
    multi: node.multi,
    label: node.label,
    transposed: node.transposed,
    filhos: node.filhos.map(c => deepCloneWithNewIds(c)),
  };
}

/**
 * When an X node has multi > 1 (N identical adjacent columns), expand it into a
 * single X node of width = valor × N, and push the repetition down to the Z level.
 *
 * Example: X680 (x2) > Y1014 > Z680
 *        → X1360      > Y1014 > Z680 (x2)
 *
 * Benefit: the space above Y1014 becomes the full 1360mm wide, enabling larger
 * pieces to fill it instead of two isolated 680mm compartments.
 */
function expandXMultiToZ(root: TreeNode): void {
  for (const xNode of root.filhos) {
    if (xNode.multi <= 1) continue;
    const N = xNode.multi;
    xNode.valor = xNode.valor * N;
    xNode.multi = 1;
    for (const yNode of xNode.filhos) {
      const originalZ = yNode.filhos.slice();
      yNode.filhos = [];
      for (let i = 0; i < N; i++) {
        for (const z of originalZ) {
          yNode.filhos.push(i === 0 ? z : deepCloneWithNewIds(z));
        }
      }
    }
  }
}

function nodesStructurallyEqual(a: TreeNode, b: TreeNode): boolean {
  if (a.tipo !== b.tipo || Math.abs(a.valor - b.valor) > 0.5) return false;
  if (a.multi !== b.multi) return false;
  if (a.label !== b.label) return false;
  if (a.filhos.length !== b.filhos.length) return false;
  for (let i = 0; i < a.filhos.length; i++) {
    if (!nodesStructurallyEqual(a.filhos[i], b.filhos[i])) return false;
  }
  return true;
}

function compressMulti(node: TreeNode): void {
  for (const child of node.filhos) {
    compressMulti(child);
  }

  if (node.filhos.length < 2) return;

  const compressed: TreeNode[] = [];
  for (const child of node.filhos) {
    if (compressed.length > 0) {
      const last = compressed[compressed.length - 1];
      if (nodesStructurallyEqual(last, child)) {
        last.multi += child.multi;
        continue;
      }
    }
    compressed.push(child);
  }
  node.filhos = compressed;
}

export function normalizeTree(tree: TreeNode, usableW: number, usableH: number, minBreak = 0): TreeNode {
  const rects = extractAbsoluteRects(tree, usableW, usableH);

  if (rects.length === 0) return tree;

  const canonical = buildCanonicalTree(rects, usableW, usableH, minBreak);
  compressMulti(canonical);
  expandXMultiToZ(canonical);
  compressMulti(canonical);
  canonical.transposed = false;

  return canonical;
}

File: wasm-engine/src/normalization.rs
use crate::types::{Arena, NodeType, NodeData, ROOT_ID, NO_PARENT};
use crate::tree_utils::children_sum;

#[derive(Clone, Debug)]
struct AbsRect {
    x: f64,
    y: f64,
    w: f64,
    h: f64,
    label: Option<String>,
}

fn extract_abs_rects(arena: &Arena, usable_w: f64, usable_h: f64) -> Vec<AbsRect> {
    let mut rects = Vec::new();
    let transposed = arena.get(ROOT_ID).transposed;
    let mut x_off = 0.0f64;

    for &x_id in &arena.nodes[ROOT_ID as usize].children.clone() {
        let x = arena.get(x_id);
        for _ in 0..x.multi {
            let mut y_off = 0.0f64;
            for &y_id in &x.children.clone() {
                let y = arena.get(y_id);
                for _ in 0..y.multi {
                    let mut z_off = 0.0f64;
                    for &z_id in &y.children.clone() {
                        let z = arena.get(z_id);
                        for _ in 0..z.multi {
                            if z.children.is_empty() {
                                rects.push(if transposed {
                                    AbsRect { x: y_off, y: x_off, w: y.valor, h: z.valor, label: z.label.clone() }
                                } else {
                                    AbsRect { x: x_off + z_off, y: y_off, w: z.valor, h: y.valor, label: z.label.clone() }
                                });
                            } else {
                                let mut w_off = 0.0f64;
                                for &w_id in &z.children.clone() {
                                    let w = arena.get(w_id);
                                    for _ in 0..w.multi {
                                        if w.children.is_empty() {
                                            rects.push(if transposed {
                                                AbsRect { x: y_off + w_off, y: x_off + z_off, w: w.valor, h: z.valor, label: w.label.clone() }
                                            } else {
                                                AbsRect { x: x_off + z_off, y: y_off + w_off, w: z.valor, h: w.valor, label: w.label.clone() }
                                            });
                                        } else {
                                            let mut q_off = 0.0f64;
                                            for &q_id in &w.children.clone() {
                                                let q = arena.get(q_id);
                                                for _ in 0..q.multi {
                                                    if q.children.is_empty() {
                                                        rects.push(if transposed {
                                                            AbsRect { x: y_off + w_off, y: x_off + z_off + q_off, w: w.valor, h: q.valor, label: q.label.clone() }
                                                        } else {
                                                            AbsRect { x: x_off + z_off + q_off, y: y_off + w_off, w: q.valor, h: w.valor, label: q.label.clone() }
                                                        });
                                                    } else {
                                                        let mut r_off = 0.0f64;
                                                        for &r_id in &q.children.clone() {
                                                            let r = arena.get(r_id);
                                                            for _ in 0..r.multi {
                                                                rects.push(if transposed {
                                                                    AbsRect { x: y_off + w_off + r_off, y: x_off + z_off + q_off, w: r.valor, h: q.valor, label: r.label.clone() }
                                                                } else {
                                                                    AbsRect { x: x_off + z_off + q_off, y: y_off + w_off + r_off, w: q.valor, h: r.valor, label: r.label.clone() }
                                                                });
                                                                r_off += r.valor;
                                                            }
                                                        }
                                                    }
                                                    q_off += q.valor;
                                                }
                                            }
                                        }
                                        w_off += w.valor;
                                    }
                                }
                            }
                            z_off += z.valor;
                        }
                    }
                    y_off += y.valor;
                }
            }
            x_off += x.valor;
        }
    }
    rects
}

fn enforce_cut_min_break(cuts: &mut Vec<f64>, bound: f64, min_break: f64) {
    if min_break <= 0.0 || cuts.is_empty() { return; }
    cuts.sort_by(|a, b| a.partial_cmp(b).unwrap());
    let mut changed = true;
    while changed {
        changed = false;
        let mut boundaries = vec![0.0];
        boundaries.extend_from_slice(cuts);
        boundaries.push(bound);
        boundaries.dedup();
        boundaries.sort_by(|a, b| a.partial_cmp(b).unwrap());
        for i in 0..boundaries.len() - 1 {
            let gap = boundaries[i + 1] - boundaries[i];
            if gap > 0.0 && gap < min_break {
                if boundaries[i + 1] < bound - 0.5 {
                    let val = boundaries[i + 1];
                    cuts.retain(|&c| (c - val).abs() > 0.5);
                } else {
                    let val = boundaries[i];
                    cuts.retain(|&c| (c - val).abs() > 0.5);
                }
                changed = true;
                break;
            }
        }
    }
}

fn find_vertical_cuts(rects: &[AbsRect], bx: f64, _by: f64, bw: f64, _bh: f64, min_break: f64) -> Vec<f64> {
    let mut edges = std::collections::HashSet::new();
    for r in rects {
        let left = (r.x - bx).round();
        let right = (r.x + r.w - bx).round();
        if left > 0.5 && left < bw - 0.5 { edges.insert(left as i64); }
        if right > 0.5 && right < bw - 0.5 { edges.insert(right as i64); }
    }

    let mut valid: Vec<f64> = edges.iter()
        .map(|&e| e as f64)
        .filter(|&cx| {
            let abs_cx = bx + cx;
            !rects.iter().any(|r| r.x < abs_cx - 0.5 && r.x + r.w > abs_cx + 0.5)
        })
        .collect();
    valid.sort_by(|a, b| a.partial_cmp(b).unwrap());
    enforce_cut_min_break(&mut valid, bw.round(), min_break);
    valid
}

fn find_horizontal_cuts(rects: &[AbsRect], _bx: f64, by: f64, _bw: f64, bh: f64, min_break: f64) -> Vec<f64> {
    let mut edges = std::collections::HashSet::new();
    for r in rects {
        let bottom = (r.y - by).round();
        let top = (r.y + r.h - by).round();
        if bottom > 0.5 && bottom < bh - 0.5 { edges.insert(bottom as i64); }
        if top > 0.5 && top < bh - 0.5 { edges.insert(top as i64); }
    }

    let mut valid: Vec<f64> = edges.iter()
        .map(|&e| e as f64)
        .filter(|&cy| {
            let abs_cy = by + cy;
            !rects.iter().any(|r| r.y < abs_cy - 0.5 && r.y + r.h > abs_cy + 0.5)
        })
        .collect();
    valid.sort_by(|a, b| a.partial_cmp(b).unwrap());
    enforce_cut_min_break(&mut valid, bh.round(), min_break);
    valid
}

fn rects_in_bounds(rects: &[AbsRect], bx: f64, by: f64, bw: f64, bh: f64) -> Vec<AbsRect> {
    rects.iter().filter(|r|
        r.x >= bx - 0.5 && r.x + r.w <= bx + bw + 0.5 &&
        r.y >= by - 0.5 && r.y + r.h <= by + bh + 0.5
    ).cloned().collect()
}

fn build_canonical_tree(rects: &[AbsRect], usable_w: f64, usable_h: f64, min_break: f64) -> Arena {
    let mut arena = Arena::new_root(usable_w);
    if rects.is_empty() { return arena; }

    let level_seq = ["X", "Y", "Z", "W", "Q", "R"];
    let is_vertical = |level: &str| matches!(level, "X" | "Z" | "Q");

    subdivide(&mut arena, ROOT_ID, 0, rects, 0.0, 0.0, usable_w, usable_h, min_break, &level_seq);
    arena
}

fn subdivide(
    arena: &mut Arena,
    parent_id: u32,
    level_idx: usize,
    sub_rects: &[AbsRect],
    bx: f64, by: f64, bw: f64, bh: f64,
    min_break: f64,
    level_seq: &[&str],
) {
    if sub_rects.is_empty() || level_idx >= level_seq.len() { return; }

    let level = level_seq[level_idx];
    let vertical = matches!(level, "X" | "Z" | "Q");

    let cuts = if vertical {
        find_vertical_cuts(sub_rects, bx, by, bw, bh, min_break)
    } else {
        find_horizontal_cuts(sub_rects, bx, by, bw, bh, min_break)
    };

    if cuts.is_empty() {
        if sub_rects.len() == 1 {
            let r = &sub_rects[0];
            let (snap_w, snap_h) = (r.w, r.h);
            if vertical {
                let nid = arena.nodes.len() as u32;
                arena.nodes.push(NodeData { tipo: NodeType::from_str(level), valor: snap_w.round(), multi: 1, children: Vec::new(), parent: parent_id, label: r.label.clone(), transposed: false });
                arena.nodes[parent_id as usize].children.push(nid);
                let needs_child = level_idx + 1 < level_seq.len() && (level_idx < 4 || snap_h.round() < bh.round());
                if needs_child {
                    let next = level_seq[level_idx + 1];
                    let child_id = arena.nodes.len() as u32;
                    arena.nodes.push(NodeData { tipo: NodeType::from_str(next), valor: snap_h.round(), multi: 1, children: Vec::new(), parent: nid, label: r.label.clone(), transposed: false });
                    arena.nodes[nid as usize].children.push(child_id);
                }
            } else {
                let nid = arena.nodes.len() as u32;
                arena.nodes.push(NodeData { tipo: NodeType::from_str(level), valor: snap_h.round(), multi: 1, children: Vec::new(), parent: parent_id, label: r.label.clone(), transposed: false });
                arena.nodes[parent_id as usize].children.push(nid);
                let needs_child = level_idx + 1 < level_seq.len() && (level_idx < 3 || snap_w.round() < bw.round());
                if needs_child {
                    let next = level_seq[level_idx + 1];
                    let child_id = arena.nodes.len() as u32;
                    arena.nodes.push(NodeData { tipo: NodeType::from_str(next), valor: snap_w.round(), multi: 1, children: Vec::new(), parent: nid, label: r.label.clone(), transposed: false });
                    arena.nodes[nid as usize].children.push(child_id);
                }
            }
            return;
        }

        if level == "X" {
            let xid = arena.nodes.len() as u32;
            arena.nodes.push(NodeData { tipo: NodeType::X, valor: bw.round(), multi: 1, children: Vec::new(), parent: parent_id, label: None, transposed: false });
            arena.nodes[parent_id as usize].children.push(xid);
            subdivide(arena, xid, level_idx + 1, sub_rects, bx, by, bw, bh, min_break, level_seq);
            return;
        }

        subdivide(arena, parent_id, level_idx + 1, sub_rects, bx, by, bw, bh, min_break, level_seq);
        return;
    }

    let bound = if vertical { bw.round() } else { bh.round() };
    let mut boundaries = vec![0.0f64];
    boundaries.extend_from_slice(&cuts);
    boundaries.push(bound);
    boundaries.dedup();
    boundaries.sort_by(|a, b| a.partial_cmp(b).unwrap());

    for i in 0..boundaries.len() - 1 {
        let seg_start = boundaries[i];
        let seg_end = boundaries[i + 1];
        let seg_size = seg_end - seg_start;
        if seg_size < 1.0 { continue; }

        let (seg_bx, seg_by, seg_bw, seg_bh) = if vertical {
            (bx + seg_start, by, seg_size, bh)
        } else {
            (bx, by + seg_start, bw, seg_size)
        };

        let seg_rects = rects_in_bounds(sub_rects, seg_bx, seg_by, seg_bw, seg_bh);
        if seg_rects.is_empty() { continue; }

        let node_valor = if vertical { seg_bw } else { seg_bh };
        let nid = arena.nodes.len() as u32;
        arena.nodes.push(NodeData {
            tipo: NodeType::from_str(level), valor: node_valor.round(), multi: 1,
            children: Vec::new(), parent: parent_id, label: None, transposed: false,
        });
        arena.nodes[parent_id as usize].children.push(nid);

        if seg_rects.len() == 1 {
            let r = &seg_rects[0];
            let fills_w = (r.w - seg_bw).abs() < 1.0;
            let fills_h = (r.h - seg_bh).abs() < 1.0;
            if fills_w && fills_h {
                arena.get_mut(nid).label = r.label.clone();
                if level_idx + 1 < level_seq.len() && level_idx < 3 {
                    let next = level_seq[level_idx + 1];
                    let child_valor = if vertical { seg_bh.round() } else { seg_bw.round() };
                    let child_id = arena.nodes.len() as u32;
                    arena.nodes.push(NodeData { tipo: NodeType::from_str(next), valor: child_valor, multi: 1, children: Vec::new(), parent: nid, label: r.label.clone(), transposed: false });
                    arena.nodes[nid as usize].children.push(child_id);
                }
                continue;
            }
        }

        subdivide(arena, nid, level_idx + 1, &seg_rects, seg_bx, seg_by, seg_bw, seg_bh, min_break, level_seq);
    }
}

fn compress_multi(arena: &mut Arena, node_id: u32) {
    let children: Vec<u32> = arena.get(node_id).children.clone();
    for &c in &children { compress_multi(arena, c); }

    if arena.get(node_id).children.len() < 2 { return; }

    let child_ids: Vec<u32> = arena.get(node_id).children.clone();
    let mut compressed: Vec<u32> = Vec::new();

    for &c in &child_ids {
        if let Some(&last_id) = compressed.last() {
            if nodes_structurally_equal(arena, last_id, c) {
                let add_multi = arena.get(c).multi;
                arena.get_mut(last_id).multi += add_multi;
                continue;
            }
        }
        compressed.push(c);
    }
    arena.get_mut(node_id).children = compressed;
}

fn nodes_structurally_equal(arena: &Arena, a: u32, b: u32) -> bool {
    let na = arena.get(a);
    let nb = arena.get(b);
    if na.tipo != nb.tipo { return false; }
    if (na.valor - nb.valor).abs() > 0.5 { return false; }
    if na.multi != nb.multi { return false; }
    if na.label != nb.label { return false; }
    if na.children.len() != nb.children.len() { return false; }
    for (&ca, &cb) in na.children.iter().zip(nb.children.iter()) {
        if !nodes_structurally_equal(arena, ca, cb) { return false; }
    }
    true
}

fn deep_clone_subtree(src: &Arena, dst: &mut Arena, src_id: u32, dst_parent: u32) -> u32 {
    let src_node = src.get(src_id);
    let new_id = dst.nodes.len() as u32;
    dst.nodes.push(NodeData {
        tipo: src_node.tipo.clone(),
        valor: src_node.valor,
        multi: src_node.multi,
        children: Vec::new(),
        parent: dst_parent,
        label: src_node.label.clone(),
        transposed: src_node.transposed,
    });
    dst.nodes[dst_parent as usize].children.push(new_id);
    let src_children: Vec<u32> = src.get(src_id).children.clone();
    for &sc in &src_children {
        deep_clone_subtree(src, dst, sc, new_id);
    }
    new_id
}

fn expand_x_multi_to_z(arena: &mut Arena) {
    let x_ids: Vec<u32> = arena.nodes[ROOT_ID as usize].children.clone();
    for x_id in x_ids {
        let n_multi = arena.get(x_id).multi;
        if n_multi <= 1 { continue; }

        let x_valor = arena.get(x_id).valor;
        arena.get_mut(x_id).valor = x_valor * n_multi as f64;
        arena.get_mut(x_id).multi = 1;

        let y_ids: Vec<u32> = arena.get(x_id).children.clone();
        for y_id in y_ids {
            let original_z: Vec<u32> = arena.get(y_id).children.clone();
            arena.get_mut(y_id).children.clear();

            // First copy: use original Z nodes
            for &z_id in &original_z {
                arena.get_mut(y_id).children.push(z_id);
            }

            // Subsequent copies: clone each Z subtree
            for _rep in 1..n_multi {
                // We need to clone from an immutable view - use the original IDs
                let clone_src_ids: Vec<u32> = original_z.clone();
                for &z_src in &clone_src_ids {
                    // Clone the Z subtree
                    let new_z_id = arena.nodes.len() as u32;
                    let src = arena.get(z_src).clone();
                    arena.nodes.push(NodeData {
                        tipo: src.tipo,
                        valor: src.valor,
                        multi: src.multi,
                        children: Vec::new(),
                        parent: y_id,
                        label: src.label,
                        transposed: src.transposed,
                    });
                    arena.nodes[y_id as usize].children.push(new_z_id);
                    // Clone children recursively
                    let src_children: Vec<u32> = arena.get(z_src).children.clone();
                    for &sc in &src_children {
                        clone_subtree_into(arena, sc, new_z_id);
                    }
                }
            }
        }
    }
}

fn clone_subtree_into(arena: &mut Arena, src_id: u32, dst_parent: u32) {
    let src = arena.get(src_id).clone();
    let new_id = arena.nodes.len() as u32;
    arena.nodes.push(NodeData {
        tipo: src.tipo, valor: src.valor, multi: src.multi,
        children: Vec::new(), parent: dst_parent,
        label: src.label, transposed: src.transposed,
    });
    arena.nodes[dst_parent as usize].children.push(new_id);
    let src_children: Vec<u32> = arena.get(src_id).children.clone();
    for &sc in &src_children {
        clone_subtree_into(arena, sc, new_id);
    }
}

pub fn normalize_tree(arena: Arena, usable_w: f64, usable_h: f64, min_break: f64) -> Arena {
    let rects = extract_abs_rects(&arena, usable_w, usable_h);
    if rects.is_empty() { return arena; }

    let mut canonical = build_canonical_tree(&rects, usable_w, usable_h, min_break);
    compress_multi(&mut canonical, ROOT_ID);
    expand_x_multi_to_z(&mut canonical);
    compress_multi(&mut canonical, ROOT_ID);
    canonical.get_mut(ROOT_ID).transposed = false;

    // Extend last X column to absorb residual
    let x_ids: Vec<u32> = canonical.nodes[ROOT_ID as usize].children.clone();
    if !x_ids.is_empty() {
        let used_w: f64 = x_ids.iter().map(|&x| { let n = canonical.get(x); n.valor * n.multi as f64 }).sum();
        let residual = (usable_w.round() - used_w.round()).max(0.0);
        if residual > 0.0 {
            let last_x = *x_ids.last().unwrap();
            if canonical.get(last_x).multi == 1 {
                canonical.get_mut(last_x).valor += residual;
            }
        }
    }

    canonical
}
